include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -0,0 +1,173 @@
+         */
+        void
+        sample (const OutputType &starting_point,
+                const std::function<double (const OutputType &)> &log_likelihood,
+                const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+                const unsigned int max_delays,
+                const types::sample_index n_samples);
+    };
+
+
+    template <typename OutputType>
+    void
+    DelayedRejectionMetropolisHastings<OutputType>::
+    sample (const OutputType &starting_point,
+            const std::function<double (const OutputType &)> &log_likelihood,
+            const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+            const unsigned int max_delays,
+            const types::sample_index n_samples)
+    {
+      // Make sure the flush_consumers() function is called at any point
+      // where we exit the current function.
+      Utilities::ScopeExit scope_exit ([this]()
+      {
+        this->flush_consumers();
+      });
+
+      std::mt19937 rng;
+      std::uniform_real_distribution<> uniform_distribution(0,1);
+
+      OutputType current_sample         = starting_point;
+      double     current_log_likelihood = log_likelihood (current_sample);
+
+      // Loop over the desired number of samples
+      for (types::sample_index i=0; i<n_samples; ++i)
+        {
+          // Initialize a vector to store rejected samples
+          std::vector<OutputType> rejected_samples;
+          // Initialize a bool to store whether a sample is accepted
+          bool accepted_sample = false;
+          // Initialize a double to store the previous acceptance_ratio;
+          // this is multiplied to the current acceptance_ratio
+          double prev_acceptance_ratio = 1.0;
+          // Delayed rejection loop
+          for (int delay_stage = 0; delay_stage <= max_delays; ++delay_stage)
+            {
+              // Obtain a new sample by perturbation and evaluate its log likelihood
+              std::pair<OutputType,double> trial_sample_and_ratio = perturb(current_sample, rejected_samples);
+              OutputType trial_sample = std::move(trial_sample_and_ratio.first);
+              const double proposal_distribution_ratio = trial_sample_and_ratio.second;
+              const double trial_log_likelihood = log_likelihood (trial_sample);
+              // Accept trial sample with probability equal to ratio of likelihoods;
+              // (always accept if > 1)
+              double acceptance_ratio = (std::exp(trial_log_likelihood - current_log_likelihood) /
+                                         proposal_distribution_ratio) * prev_acceptance_ratio;
+              double u = uniform_distribution(rng);
+              assert(acceptance_ratio > 1 == acceptance_ratio >= u);
+              if (acceptance_ratio > 1 || acceptance_ratio >= u)
+                accepted_sample = true;
```

Comment:
@bangerth This assert passes just fine, _but_ with this setup (using a variable `u`), the mean value converges to 1 instead of 2. Simply removing the variable `u` and calling `uniform_distribution()` inside the conditional results in a mean of 2.

Reply:
@bangerth Any thoughts on this? Why would the assignment of a variable affect the random values?